runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/StreamingEngineClient.java

```diff
@@ -0,0 +1,432 @@
+            .setGlobalDataEndpoints(newWindmillEndpoints.globalDataEndpoints())
+            .setGlobalDataStreams(newGlobalDataStreams)
+            .build();
+
+    LOG.info(
+        "Setting new connections: {}. Previous connections: {}.",
+        newConnectionsState,
+        connections.get());
+    connections.set(newConnectionsState);
+    isBudgetRefreshPaused.set(false);
+
+    // Trigger on first worker metadata. Since startAndCacheStreams will block until first worker
+    // metadata is ready. Afterwards, just trigger a budget refresh.
+    if (getWorkerMetadataReady.getCount() > 0) {
+      getWorkerMetadataReady.countDown();
+    } else {
+      getWorkBudgetRefresher.requestBudgetRefresh();
+    }
+  }
+
+  public final ImmutableList<Long> getAndResetThrottleTimes() {
+    StreamEngineConnectionState currentConnections = connections.get();
+
+    ImmutableList<Long> keyedWorkStreamThrottleTimes =
+        currentConnections.windmillStreams().values().stream()
+            .map(WindmillStreamSender::getAndResetThrottleTime)
+            .collect(toImmutableList());
+
+    return ImmutableList.<Long>builder()
+        .add(getWorkerMetadataThrottleTimer.getAndResetThrottleTime())
+        .addAll(keyedWorkStreamThrottleTimes)
+        .build();
+  }
+
+  /** Starts {@link GetWorkerMetadataStream}. */
+  @SuppressWarnings({
+    "FutureReturnValueIgnored", // ignoring Future returned from Executor.submit()
+    "nullness" // Uninitialized value of getWorkerMetadataStream is null.
+  })
+  private void startGetWorkerMetadataStream() {
+    // We only want to set and start this value once.
+    if (getWorkerMetadataStream == null) {
+      synchronized (this) {
+        if (getWorkerMetadataStream == null) {
+          getWorkerMetadataStream =
+              streamFactory.createGetWorkerMetadataStream(
+                  dispatcherClient.getDispatcherStub(),
+                  getWorkerMetadataThrottleTimer,
+                  endpoints ->
+                      // Run this on a separate thread than the grpc stream thread.
+                      consumeWorkerMetadataExecutor.submit(
+                          () -> consumeWindmillWorkerEndpoints(endpoints)));
+        }
+      }
+    }
+  }
+
+  private synchronized ImmutableMap<Endpoint, WindmillConnection> createNewWindmillConnections(
+      ImmutableSet<Endpoint> newWindmillEndpoints) {
+    ImmutableMap<Endpoint, WindmillConnection> currentWindmillConnections =
+        connections.get().windmillConnections();
+    Map<Endpoint, WindmillConnection> newWindmillConnections =
+        currentWindmillConnections.entrySet().stream()
+            .filter(entry -> newWindmillEndpoints.contains(entry.getKey()))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    // Reuse existing stubs if they exist.
+    newWindmillEndpoints.forEach(
+        newEndpoint ->
+            newWindmillConnections.putIfAbsent(
```

Comment:
how about just a single loop doing this

newWindmillConnections.put(
   newEndpoint, 
   currentWindmillConnections.getOrDefault(
      newEndpoint, WindmillConnection.from(newEndpoint, this::createWindmillStub)));

still constructs WindmillConnection even if not necessary but that isn't expensive.

Reply:
done